docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,150 @@
+- It makes it difficult to manage different Kubelet configurations for different nodes, e.g. if you want to canary a new config.
+- The current lack of versioned Kubelet configuration means that any time we change Kubelet flags, we risk breaking someone's setup.
+- Having the ability to pass config in the same format on-disk as via the API server will make it easier to bootstrap master nodes before the API server is available.
+
+## Example Use Cases
+- Configuring heterogeneous machines in a cluster, where machines need different node-allocatable settings, different eviction settings, etc.
+- Configuring nodes with different container runtimes in the same cluster.
+- Staged rollout of new Kubelet features. This feature will allow you to configure a few nodes at a time with new feature enabled, before rolling out that configuration to the entire fleet.
+
+
+## Goals of this design:
+
+- Add a well-defined, versioned configmap object for Kubelet configuration.
+- Add/Remove/Update per-node Kubelet config on the fly via the API server.
+- Provide a reliable way to get the configuration of a given Kubelet along with its source (remote or on disk or flags).
+- We want the Kubelet to work with the built-in defaults on all distros we test against.
+- Work towards deprecating flags in favor of on-disk config and config from the API server. Master nodes will be bootstrapped from on-disk config.
+- Make it possible to opt-out of remote configuration e.g. to protect the master from bad config updates.
+
+## Design
+
+### Definitions
+
+- `kubelet-default`: The built-in default config that every Kubelet starts up with.
+- `api-default`: The defaults provided for omitted/nil fields by the API server.
+- `on-disk-config`: `KubeletConfiguration` provided on disk, used as a fallback when the Kubelet cannot contact the API server.
+- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via `on-disk-config`.
+- `kubelet-<node-name>`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, set via the API server. **Note**: I changed this from `<node-name>-kubelet-configuration` so that an alphabetical sort groups by purpose (config), then by component (kubelet), then by node. It doesn't seem like we've really established configmap naming conventions, but I think this could be a reasonable strategy. (It could also be reasonable to do `component-purpose-<node-name>`, IMO). 
+- **Note**: We've decided against using any sort of "global" Kubelet configuration object. Kubelets try to take up config immediately upon seeing new config, which could be very dangerous in the event of a bad config push. It is better to have a configmap for each node, and to have to think about a rollout strategy for updating config for groups of machines.
```

Comment:
Using node names as our unit of roll-out seems to be treating the nodes as pets, not cattle, and doesn't feel very consistent with the way that k8s does other things.

Have we considered instead using labels & selectors?  That is consistent with k8s, permits the same granularity that per-node selection does, but also enables higher level management.

For example, on AWS, node names are not very meaningful - nodes come and go.  We might even choose to implement a kubelet reconfiguration by replacing instances, in which case this is going to be really problematic, because the node name isn't known until the new node starts up.